substrate/runtime/primitives/src/lib.rs

```diff
@@ -55,24 +62,88 @@ pub trait BuildExternalities {
 
 /// Ed25519 signature verify.
 #[derive(Eq, PartialEq, Clone, Default)]
-#[cfg_attr(feature = "std", derive(Debug, Serialize))]
-pub struct Ed25519Signature(H512);
-impl traits::Verify for Ed25519Signature {
-	type Signer = [u8; 32];
-	fn verify(&self, msg: &[u8], signer: &Self::Signer) -> bool {
-		runtime_io::ed25519_verify(&(self.0).0, msg, &signer[..])
+#[cfg_attr(feature = "std", derive(Debug, Serialize, Deserialize))]
+pub struct Ed25519Signature(pub H512);
+
+impl Verify for Ed25519Signature {
+	type Signer = H256;
+	fn verify<L: Lazy<[u8]>>(&self, mut msg: L, signer: &Self::Signer) -> bool {
+		runtime_io::ed25519_verify(&(self.0).0, msg.get(), &signer.0[..])
 	}
 }
+
 impl codec::Slicable for Ed25519Signature {
 	fn decode<I: codec::Input>(input: &mut I) -> Option<Self> { Some(Ed25519Signature(codec::Slicable::decode(input)?,)) }
 	fn using_encoded<R, F: FnOnce(&[u8]) -> R>(&self, f: F) -> R { self.0.using_encoded(f) }
 }
+
 impl From<H512> for Ed25519Signature {
 	fn from(h: H512) -> Ed25519Signature {
 		Ed25519Signature(h)
 	}
 }
 
+/// Potentially "unsigned" signature verification.
+#[derive(Eq, PartialEq, Clone, Default)]
+#[cfg_attr(feature = "std", derive(Debug, Serialize, Deserialize))]
+pub struct MaybeUnsigned<T>(pub T);
+
+impl<T: Verify> MaybeUnsigned<T> where
+	T: Default + Eq,
+	<T as Verify>::Signer: Default + Eq,
+{
+	fn is_signed(&self, signer: &<Self as Verify>::Signer) -> bool {
+		self.0 != T::default() || signer != &<Self as Verify>::Signer::default()
+	}
+}
+
+impl<T: Verify> Verify for MaybeUnsigned<T> where
+	T: Default + Eq,
+	<T as Verify>::Signer: Default + Eq,
+{
+	type Signer = T::Signer;
+	fn verify<L: Lazy<[u8]>>(&self, msg: L, signer: &Self::Signer) -> bool {
+		if !self.is_signed(signer) {
+			true
+		} else {
+			self.0.verify(msg, signer)
+		}
+	}
+}
+
+impl<T: codec::Slicable> codec::Slicable for MaybeUnsigned<T> {
+	fn decode<I: codec::Input>(input: &mut I) -> Option<Self> { Some(MaybeUnsigned(codec::Slicable::decode(input)?)) }
+	fn using_encoded<R, F: FnOnce(&[u8]) -> R>(&self, f: F) -> R { self.0.using_encoded(f) }
+}
+
+impl<T> From<T> for MaybeUnsigned<T> {
+	fn from(t: T) -> Self {
+		MaybeUnsigned(t)
+	}
+}
+
+/// Verify a signature on an encoded value in a lazy manner. This can be
+/// an optimization if the signature scheme has an "unsigned" escape hash.
+pub fn verify_encoded_lazy<V: Verify, T: codec::Slicable>(sig: &V, item: &T, signer: &V::Signer) -> bool {
+	// TODO: unfortunately this is a lifetime relationship that can't
```

Comment:
Issue link + description of how to change once issue is sorted would be good.

Reply:
we would need https://github.com/rust-lang/rust/issues/44265 and then the `Fn*` traits would have to be extended to use it somehow. I don't think it's likely to be something we can work around until a few more epochs of Rust. I'll remove the `TODO` because it's not reasonable to do in the near future.